src/services_models/services_constructor.jl

```diff
@@ -15,96 +44,182 @@ end
 function construct_services!(
     container::OptimizationContainer,
     sys::PSY.System,
+    stage::ArgumentConstructStage,
     services_template::ServicesModelContainer,
     devices_template::DevicesModelContainer,
 )
     isempty(services_template) && return
     incompatible_device_types = get_incompatible_devices(devices_template)
 
-    function _construct_valid_services!(service_model::ServiceModel)
-        @debug "Building $(get_component_type(service_model)) with $(get_formulation(service_model)) formulation"
-        services = get_component_type(service_model)[]
-        if validate_services!(
-            get_component_type(service_model),
-            services,
-            incompatible_device_types,
+    groupservice = nothing
+
+    for (key, service_model) in services_template
```

Comment:
do we need this type check at run time?

Reply:
I just followed the existing code logic to build the group reserves last, we could make the `ServicesModelContainer` a OrderedDict to make sure the group reserve is always the last.